pipeline/schema/schema.json

```diff
@@ -304,13 +258,118 @@
+      ]
+    },
+    "step_environment": {
+      "description": "Pass environment variables to a pipeline step. Read more: https://woodpecker-ci.org/docs/usage/environment",
+      "oneOf": [
+        {
+          "type": "array",
+          "items": { "type": "string" },
+          "minLength": 1
+        },
+        {
+          "type": "object",
+          "additionalProperties": {
+            "type": ["boolean", "string", "number"]
+          }
+        }
+      ]
+    },
+    "step_secrets": {
+      "description": "Pass secrets to a pipeline step at runtime. Read more: https://woodpecker-ci.org/docs/usage/secrets",
+      "type": "array",
+      "items": {
+        "oneOf": [
+          { "type": "string" },
+          {
+            "type": "object",
+            "required": ["source", "target"],
+            "properties": {
+              "source": { "type": "string" },
+              "target": { "type": "string" }
+            }
+          }
+        ]
+      },
+      "minLength": 1
+    },
+    "step_settings": {
+      "description": "Change the settings of your plugin. Read more: https://woodpecker-ci.org/docs/usage/plugins/plugins",
+      "type": "object",
+      "additionalProperties": {
+        "type": ["boolean", "string", "number", "array", "object"]
+      }
+    },
+    "step_volumes": {
+      "description": "Mount files or folders from the host machine into your step container. Read more: https://woodpecker-ci.org/docs/usage/volumes",
+      "oneOf": [
+        { "type": "string" },
+        { "type": "array", "items": { "type": "string" }, "minLength": 1 }
+      ]
+    },
     "services": {
-      "description": "TODO Read more: https://woodpecker-ci.org/docs/usage/services",
+      "description": "Read more: https://woodpecker-ci.org/docs/usage/services",
       "type": "object",
-      "additionalProperties": true
+      "additionalProperties": { "$ref": "#/definitions/service" },
+      "minProperties": 1
+    },
+    "service": {
+      "description": "Read more: https://woodpecker-ci.org/docs/usage/pipeline-syntax#steps",
+      "type": "object",
+      "additionalProperties": false,
+      "minProperties": 1,
+      "required": ["image"],
+      "properties": {
+        "image": {
+          "$ref": "#/definitions/step_image"
+        },
+        "pull": {
+          "$ref": "#/definitions/step_pull"
+        },
+        "commands": {
+          "$ref": "#/definitions/step_commands"
+        },
+        "environment": {
+          "$ref": "#/definitions/step_environment"
+        },
+        "secrets": {
+          "$ref": "#/definitions/step_secrets"
+        },
+        "settings": {
+          "$ref": "#/definitions/step_settings"
+        },
+        "when": {
+          "$ref": "#/definitions/step_when"
+        },
+        "volumes": {
+          "$ref": "#/definitions/step_volumes"
+        },
+        "ports": {
+          "description": "export port steps in pipeline can connect to",
```

Comment:
```suggestion
          "description": "export ports, witch steps can connect to",
```
?

Reply:
-> I have to add it to the schema or our pipeline linting step will fail - wo do use it :sweat_smile: